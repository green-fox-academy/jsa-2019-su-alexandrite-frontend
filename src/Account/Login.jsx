import React, { useState, useEffect } from 'react';
import {
  TextInput,
  View,
  KeyboardAvoidingView,
  TouchableHighlight,
  Text,
  ActivityIndicator,
} from 'react-native';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigation } from 'react-navigation-hooks';
import { FontAwesome5 } from '@expo/vector-icons';
import PropTypes from 'prop-types';
import { loginUser } from '../redux/account/actionCreator';
import styles from './styles';
import Row from '../common/Row';
import Column from '../common/Column';

const LoginButton = ({ onPress, isLoading }) => (
  <TouchableHighlight style={styles.button} onPress={onPress} underlayColor="#5d70ba" activeOpacity={0.5}>
    {isLoading
      ? <ActivityIndicator size="large" color="white" />
      : (
        <Text style={styles.buttonText}>
          <FontAwesome5 name="arrow-right" size={16} />
        </Text>
      )}
  </TouchableHighlight>
);

const Login = () => {
  const [username, setUsername] = useState('');
  const [password, setPassword] = useState('');
  const dispatch = useDispatch();
  const { pop } = useNavigation();

  const { accessToken, error, isLoggingIn } = useSelector((state) => state.user);

  useEffect(() => {
    if (accessToken) pop();
  }, [accessToken]);

  const login = () => {
    dispatch(loginUser(username, password));
  };

  return (
    <KeyboardAvoidingView style={styles.container} behavior="padding" enabled>
      <Row style={{ justifyContent: 'center', alignItems: 'center' }}>
        <Text style={styles.loginTitle}>Login</Text>
      </Row>
      <Row style={{ flex: 2 }}>
        <View style={styles.inputContainer}>
          <TextInput
            placeholder="Username"
            label="Username"
            name="username"
            autoCapitalize="none"
            autoCorrect={false}
            style={styles.textInput}
            onChangeText={(text) => setUsername(text)}
            value={username}
          />
          <TextInput
            secureTextEntry
            placeholder="Password"
            label="Password"
            name="password"
            autoCapitalize="none"
            autoCorrect={false}
            style={styles.textInput}
            type="password"
            onChangeText={(text) => setPassword(text)}
            value={password}
          />
          {error
            ? (
              <View style={styles.errorMessageContainer}>
                <Text style={styles.errorMessage}>
                  {error.message}
                </Text>
              </View>
            )
            : null}
          <Row style={{ flex: 0, justifyContent: 'center' }}>
            <Column style={{ flex: 0 }}>
              <LoginButton onPress={login} isLoading={isLoggingIn} />
            </Column>
          </Row>
        </View>
      </Row>
    </KeyboardAvoidingView>
  );
};

LoginButton.propTypes = {
  onPress: PropTypes.func.isRequired,
  isLoading: PropTypes.bool.isRequired,
};

export default Login;
