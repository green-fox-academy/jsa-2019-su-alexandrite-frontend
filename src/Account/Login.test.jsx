import React from 'react';
import renderer from 'react-test-renderer';
import { useDispatch, useSelector } from 'react-redux';
import { useNavigation } from 'react-navigation-hooks';

import Login from './Login';

jest.mock('react-redux');
jest.mock('../redux/account/actionCreator');
jest.mock('react-navigation-hooks');

const SAMPLE_STATE = {
  accseeToken: 'aaabbbcccdddeeefffggghhhiiijjjkkkklllmmmnnn',
  error: null,
  isLoggingIn: false,
};

describe('<Login />', () => {
  beforeEach(() => {
    useDispatch.mockReturnValue(jest.fn(() => { }));
  });

  it('Expect to not log errors in console', () => {
    useSelector.mockReturnValue(SAMPLE_STATE);
    const pop = jest.fn(() => { });
    useNavigation.mockReturnValue({ pop });
    const spy = jest.spyOn(global.console, 'error');
    const wrapper = renderer.create(
      <Login />,
    );
    expect(wrapper).not.toBeNull();
    expect(spy).not.toHaveBeenCalled();
  });

  it('Should render and match the snapshot without error', () => {
    useSelector.mockReturnValue(SAMPLE_STATE);
    const tree = renderer.create(
      <Login />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render and match the snapshot with error', () => {
    useSelector.mockReturnValue({ ...SAMPLE_STATE, error: 'error message' });

    const tree = renderer.create(
      <Login isLoading="true" />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('Should render and match the snapshot when loading', () => {
    useSelector.mockReturnValue({ ...SAMPLE_STATE, isLoggingIn: true });
    const tree = renderer.create(
      <Login />,
    ).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
