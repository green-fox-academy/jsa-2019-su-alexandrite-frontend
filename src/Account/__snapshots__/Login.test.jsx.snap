// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> Should render and match the snapshot when loading 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 36,
          "fontWeight": "bold",
        }
      }
    >
      Login
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 2,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        label="Username"
        name="username"
        onChangeText={[Function]}
        placeholder="Username"
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "#fff",
            "borderColor": "#eee",
            "borderRadius": 5,
            "borderWidth": 2,
            "fontSize": 16,
            "fontWeight": "bold",
            "marginHorizontal": 15,
            "marginTop": 15,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        label="Password"
        name="password"
        onChangeText={[Function]}
        placeholder="Password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "#fff",
            "borderColor": "#eee",
            "borderRadius": 5,
            "borderWidth": 2,
            "fontSize": 16,
            "fontWeight": "bold",
            "marginHorizontal": 15,
            "marginTop": 15,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
          }
        }
        type="password"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "flex": 0,
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 0,
              "flexDirection": "column",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#566ed3",
                "borderRadius": 50,
                "elevation": 3,
                "height": 60,
                "justifyContent": "center",
                "marginTop": 50,
                "shadowColor": "#566ed3",
                "shadowOffset": Object {
                  "height": 3,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 6,
                "width": 60,
              }
            }
          >
            <ActivityIndicator
              animating={true}
              color="white"
              hidesWhenStopped={true}
              size="large"
              style={null}
            />
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`<Login /> Should render and match the snapshot with error 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 36,
          "fontWeight": "bold",
        }
      }
    >
      Login
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 2,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        label="Username"
        name="username"
        onChangeText={[Function]}
        placeholder="Username"
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "#fff",
            "borderColor": "#eee",
            "borderRadius": 5,
            "borderWidth": 2,
            "fontSize": 16,
            "fontWeight": "bold",
            "marginHorizontal": 15,
            "marginTop": 15,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        label="Password"
        name="password"
        onChangeText={[Function]}
        placeholder="Password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "#fff",
            "borderColor": "#eee",
            "borderRadius": 5,
            "borderWidth": 2,
            "fontSize": 16,
            "fontWeight": "bold",
            "marginHorizontal": 15,
            "marginTop": 15,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
          }
        }
        type="password"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "alignItems": "center",
            "paddingTop": 10,
          }
        }
      >
        <Text
          style={
            Object {
              "color": "#9f3a38",
              "fontSize": 15,
            }
          }
        >
          error message
        </Text>
      </View>
      <View
        style={
          Object {
            "flex": 0,
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 0,
              "flexDirection": "column",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#566ed3",
                "borderRadius": 50,
                "elevation": 3,
                "height": 60,
                "justifyContent": "center",
                "marginTop": 50,
                "shadowColor": "#566ed3",
                "shadowOffset": Object {
                  "height": 3,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 6,
                "width": 60,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontWeight": "bold",
                }
              }
            >
              <Text />
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;

exports[`<Login /> Should render and match the snapshot without error 1`] = `
<View
  onLayout={[Function]}
  style={
    Array [
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "column",
        "justifyContent": "center",
      },
      Object {
        "paddingBottom": 0,
      },
    ]
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flex": 1,
        "flexDirection": "row",
        "justifyContent": "center",
      }
    }
  >
    <Text
      style={
        Object {
          "fontSize": 36,
          "fontWeight": "bold",
        }
      }
    >
      Login
    </Text>
  </View>
  <View
    style={
      Object {
        "flex": 2,
        "flexDirection": "row",
      }
    }
  >
    <View
      style={
        Object {
          "flex": 1,
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        label="Username"
        name="username"
        onChangeText={[Function]}
        placeholder="Username"
        rejectResponderTermination={true}
        style={
          Object {
            "backgroundColor": "#fff",
            "borderColor": "#eee",
            "borderRadius": 5,
            "borderWidth": 2,
            "fontSize": 16,
            "fontWeight": "bold",
            "marginHorizontal": 15,
            "marginTop": 15,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <TextInput
        allowFontScaling={true}
        autoCapitalize="none"
        autoCorrect={false}
        label="Password"
        name="password"
        onChangeText={[Function]}
        placeholder="Password"
        rejectResponderTermination={true}
        secureTextEntry={true}
        style={
          Object {
            "backgroundColor": "#fff",
            "borderColor": "#eee",
            "borderRadius": 5,
            "borderWidth": 2,
            "fontSize": 16,
            "fontWeight": "bold",
            "marginHorizontal": 15,
            "marginTop": 15,
            "paddingHorizontal": 15,
            "paddingVertical": 10,
          }
        }
        type="password"
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        style={
          Object {
            "flex": 0,
            "flexDirection": "row",
            "justifyContent": "center",
          }
        }
      >
        <View
          style={
            Object {
              "flex": 0,
              "flexDirection": "column",
            }
          }
        >
          <View
            accessible={true}
            focusable={true}
            isTVSelectable={true}
            onClick={[Function]}
            onResponderGrant={[Function]}
            onResponderMove={[Function]}
            onResponderRelease={[Function]}
            onResponderTerminate={[Function]}
            onResponderTerminationRequest={[Function]}
            onStartShouldSetResponder={[Function]}
            style={
              Object {
                "alignItems": "center",
                "backgroundColor": "#566ed3",
                "borderRadius": 50,
                "elevation": 3,
                "height": 60,
                "justifyContent": "center",
                "marginTop": 50,
                "shadowColor": "#566ed3",
                "shadowOffset": Object {
                  "height": 3,
                },
                "shadowOpacity": 0.3,
                "shadowRadius": 6,
                "width": 60,
              }
            }
          >
            <Text
              style={
                Object {
                  "color": "#fff",
                  "fontWeight": "bold",
                }
              }
            >
              <Text />
            </Text>
          </View>
        </View>
      </View>
    </View>
  </View>
</View>
`;
